### ISSUE TITLE
[DOC] Migrating Jupyter 2 recipes to Jupyter 3

### ISSUE THREAD

**Issue by jakirkham** [conda-forge/conda-forge.github.io#1421]
Not entirely sure what this entails personally. So am simply filing this issue to acknowledge this is something we need help with from the Jupyter community and that we then need to communicate this in the docs so maintainers can help us make these changes. Cc-ing a few people that may help provide guidance here (though feel free to include others 🙂)

cc @SylvainCorlay @jasongrout @bollwyvl

**Comment by bollwyvl** [comment]
Thanks for getting this up!

Not only 2 `->` 3, but 2 **`+`** 3 (in the same package) is a thing. Luckily I don't know of anybody _also_ trying to support 1 :scream:  !

Also, it might well get _real_ nasty soon, if `data_files` is finally [shown the door](https://discourse.jupyter.org/t/how-could-data-files-be-improved/8972/2).

I'll do some more concerted thinkifying about it this evening.

:bellhop_bell:  for SA: @blink1073

**Comment by jakirkham** [comment]
Does it make sense to build a single Conda package that supports both 2 + 3? Or is this sort of thing where we would need different Conda packages for each major version?

**Comment by bollwyvl** [comment]
it works fine, with the wrinkle: 2 "support" means "puts a tarball in a
place, and probably needs nodejs" while 3 means "put a folder in another
place"

**Comment by jakirkham** [comment]
Are there `run` dependency differences between the two builds (outside of `nodejs`)?

**Comment by bollwyvl** [comment]
> Are there `run` dependency differences between the two builds

Not really! For example, widget packages will often be independent of the lab version altogether, and are just using `pip` as a distribution mechanism for built JS assets.

I think the biggest thing is to avoid `jupyterlab` and `nodejs` in `host`/`build` altogether, as all well-behaved extensions will already contain the built assets, and you never _really_ know what you're going to get with webpack that day inside of `pip install` vs. what the upstream built (and hopefully _tested as built_).  I also _really_ hope the trend of making `jupyterlab-packaging` a `run` dependency does not propagate, as there are breaking API changes planned that are going to cripple peoples' ability to install extensions that depend on it for a while after it gets adopted, according to my :crystal_ball:... or :8ball: ...

To that end, nasty things like deleting `pyproject.toml` are sometimes required already, as `PIP_NO_BUILD_ISOLATION` doesn't always do what is says on the tin.
